libminifi/test/integration/C2MetricsTest.cpp

```diff
@@ -131,23 +134,67 @@ class MetricsHandler: public HeartbeatHandler {
       runtime_metrics.HasMember("flowInfo") &&
       runtime_metrics["flowInfo"].HasMember("versionedFlowSnapshotURI") &&
       runtime_metrics["flowInfo"].HasMember("queues") &&
-      runtime_metrics["flowInfo"].HasMember("components") &&
       runtime_metrics["flowInfo"]["queues"].HasMember("2438e3c8-015a-1000-79ca-83af40ec1997") &&
-      runtime_metrics["flowInfo"]["components"].HasMember("FlowController") &&
-      runtime_metrics["flowInfo"]["components"].HasMember("GetTCP") &&
-      runtime_metrics["flowInfo"]["components"].HasMember("LogAttribute");
+      runtime_metrics["flowInfo"].HasMember("processorStatuses") &&
+      [&]() {
+        if (runtime_metrics["flowInfo"]["processorStatuses"].GetArray().Size() != 2) {
+          return false;
+        }
+        for (const auto& processor : runtime_metrics["flowInfo"]["processorStatuses"].GetArray()) {  // NOLINT(readability-use-anyofallof)
+          if (processor["id"].GetString() != std::string(GETTCP_UUID) && processor["id"].GetString() != std::string(LOGATTRIBUTE1_UUID)) {
+            throw std::runtime_error(std::string("Unexpected processor id in processorStatuses: ") + processor["id"].GetString());
+          }
+
+          if (processor["bytesRead"].GetInt() < 0 ||
+              processor["bytesWritten"].GetInt() < 0 ||
+              processor["flowFilesIn"].GetInt() < 0 ||
+              processor["flowFilesOut"].GetInt() < 0 ||
+              processor["bytesIn"].GetInt() < 0 ||
+              processor["bytesOut"].GetInt() < 0 ||
+              processor["invocations"].GetInt() < 0 ||
+              processor["processingNanos"].GetInt() < 0 ||
+              processor["activeThreadCount"].GetInt() != -1 ||
+              processor["terminatedThreadCount"].GetInt() != -1 ||
+              !processor["running"].GetBool()) {
+            return false;
+          }
+        }
+        return true;
+      }();
   }
 
   static bool verifyUpdatedRuntimeMetrics(const rapidjson::Value& runtime_metrics) {
     return runtime_metrics.HasMember("deviceInfo") &&
       runtime_metrics.HasMember("flowInfo") &&
       runtime_metrics["flowInfo"].HasMember("versionedFlowSnapshotURI") &&
       runtime_metrics["flowInfo"].HasMember("queues") &&
-      runtime_metrics["flowInfo"].HasMember("components") &&
       runtime_metrics["flowInfo"]["queues"].HasMember("8368e3c8-015a-1003-52ca-83af40ec1332") &&
-      runtime_metrics["flowInfo"]["components"].HasMember("FlowController") &&
-      runtime_metrics["flowInfo"]["components"].HasMember("GenerateFlowFile") &&
-      runtime_metrics["flowInfo"]["components"].HasMember("LogAttribute");
+      runtime_metrics["flowInfo"].HasMember("processorStatuses") &&
+      [&]() {
+        if (runtime_metrics["flowInfo"]["processorStatuses"].GetArray().Size() != 2) {
+          return false;
+        }
+        for (const auto& processor : runtime_metrics["flowInfo"]["processorStatuses"].GetArray()) {  // NOLINT(readability-use-anyofallof)
```

Comment:
same as above. It also looks like code duplication, consider extracting the common code to a local helper function.

Reply:
Updated in https://github.com/apache/nifi-minifi-cpp/pull/1909/commits/2723bace043147136941c68f86216fcd61d74a36